broker/src/task_manager.rs

```diff
@@ -0,0 +1,329 @@
+    ops::Deref,
+    time::{Duration, SystemTime}, collections::HashMap, sync::Arc,
+};
+
+use axum::{response::{IntoResponse, sse::Event, Sse}, Json};
+use dashmap::DashMap;
+use futures_core::Stream;
+use hyper::StatusCode;
+use once_cell::sync::Lazy;
+use serde::Serialize;
+use shared::{
+    beam_id::AppOrProxyId, HasWaitId, HowLongToBlock, Msg, MsgEmpty, MsgId, MsgSigned,
+    MsgState, MsgTaskRequest, MsgTaskResult, sse_event::SseEventType,
+};
+use tokio::{sync::broadcast, time::Instant};
+use tracing::{warn, error};
+
+pub trait Task {
+    type Result;
+
+    fn get_results(&self) -> &HashMap<AppOrProxyId, Self::Result>;
+    /// Returns true if the value as been updated and false if it was a result from a new app
+    fn insert_result(&mut self, result: Self::Result) -> bool;
+    fn is_expired(&self) -> bool;
+}
+
+impl<State: MsgState> Task for MsgTaskRequest<State> {
+    type Result = MsgSigned<MsgTaskResult<State>>;
+
+    fn insert_result(&mut self, result: Self::Result) -> bool {
+        self.results.insert(result.get_from().clone(), result).is_some()
+    }
+
+    fn get_results(&self) -> &HashMap<AppOrProxyId, Self::Result> {
+        &self.results
+    }
+
+    fn is_expired(&self) -> bool {
+        self.expire < SystemTime::now()
+    }
+}
+
+static EMPTY_MAP: Lazy<HashMap<AppOrProxyId, ()>> = Lazy::new(|| {
+    HashMap::with_capacity(0)
+});
+
+#[cfg(feature = "sockets")]
+impl<State: MsgState> Task for shared::MsgSocketRequest<State> {
+    type Result = ();
+
+    fn get_results(&self) -> &HashMap<AppOrProxyId, Self::Result> {
+        &EMPTY_MAP
+    }
+
+    fn insert_result(&mut self, _result: Self::Result) -> bool { false }
+
+    fn is_expired(&self) -> bool {
+        self.expire < SystemTime::now()
+    }
+}
+
+pub struct TaskManager<T: HasWaitId<MsgId> + Task + Msg> {
+    tasks: DashMap<MsgId, MsgSigned<T>>,
+    new_tasks: broadcast::Sender<MsgId>,
+    /// Send the index at which the new result for the given Task was inserted
+    new_results: DashMap<MsgId, broadcast::Sender<AppOrProxyId>>,
+}
+
+impl<T: HasWaitId<MsgId> + Task + Msg + Send + Sync + 'static> TaskManager<T> {
+    const EXPIRE_CHECK_INTERVAL: Duration = Duration::from_secs(5 * 60);
+
+    pub fn new() -> Arc<Self> {
+        let (new_tasks, _) = broadcast::channel(256);
+        let task_manager = Arc::new(Self {
+            tasks: Default::default(),
+            new_tasks,
+            new_results: Default::default(),
+        });
+        let tm = Arc::clone(&task_manager);
+        tokio::spawn(async move {
+            loop {
+                tokio::time::sleep(Self::EXPIRE_CHECK_INTERVAL).await;
+                tm.tasks.retain(|_, task| !task.msg.is_expired());
+                // TODO: Maybe do a DashMap::shrink_to_fit here if capacity is way larger than number of elements
+                // This will need to be tested as shrinking locks the whole map making it inaccessible until
+                // everything is reallocated
```

Comment:
This is currently not required, our memory footprint is small enough, that I would love to not introduce this locking penalty. If it poses a problem, we could think about using a custom allocator.

Reply:
Will update the comment to not include a todo then.